File: packages/core/src/__tests__/VersionNegotiator.test.ts
import { describe, it, expect } from 'vitest';
import { VersionNegotiator } from '../version/VersionNegotiator.js';

describe('VersionNegotiator', () => {
  describe('parse', () => {
    it('parses valid semver', () => {
      expect(VersionNegotiator.parse('1.2.3')).toEqual({ major: 1, minor: 2, patch: 3 });
      expect(VersionNegotiator.parse('0.1.0')).toEqual({ major: 0, minor: 1, patch: 0 });
    });

    it('throws on invalid format', () => {
      expect(() => VersionNegotiator.parse('1.2')).toThrow(/Invalid version/);
      expect(() => VersionNegotiator.parse('abc')).toThrow(/Invalid version/);
      expect(() => VersionNegotiator.parse('1.2.3-beta')).toThrow(/Invalid version/);
    });
  });

  describe('isCompatible', () => {
    it('same major version is compatible (>= 1.0.0)', () => {
      expect(VersionNegotiator.isCompatible('1.0.0', '1.5.3')).toBe(true);
      expect(VersionNegotiator.isCompatible('2.0.0', '2.1.0')).toBe(true);
    });

    it('different major version is incompatible', () => {
      expect(VersionNegotiator.isCompatible('1.0.0', '2.0.0')).toBe(false);
      expect(VersionNegotiator.isCompatible('3.0.0', '1.0.0')).toBe(false);
    });

    it('0.x uses minor for compatibility', () => {
      expect(VersionNegotiator.isCompatible('0.1.0', '0.1.5')).toBe(true);
      expect(VersionNegotiator.isCompatible('0.1.0', '0.2.0')).toBe(false);
    });
  });

  describe('assertCompatible', () => {
    it('does not throw for compatible versions', () => {
      expect(() => VersionNegotiator.assertCompatible('1.0.0', '1.2.0')).not.toThrow();
    });

    it('throws for incompatible versions', () => {
      expect(() => VersionNegotiator.assertCompatible('1.0.0', '2.0.0')).toThrow(
        /incompatible/,
      );
    });
  });
});

File: android/core/src/test/kotlin/com/devryner/rynbridge/core/VersionNegotiatorTest.kt
package com.devryner.rynbridge.core

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

class VersionNegotiatorTest {

    private val negotiator = VersionNegotiator()

    @Test
    fun `parse valid version`() {
        val v = negotiator.parse("1.2.3")
        assertEquals(1, v.major)
        assertEquals(2, v.minor)
        assertEquals(3, v.patch)
    }

    @Test
    fun `parse invalid version`() {
        val exception = assertThrows(RynBridgeError::class.java) {
            negotiator.parse("invalid")
        }
        assertEquals(ErrorCode.VERSION_MISMATCH, exception.code)
    }

    @Test
    fun `parse incomplete version`() {
        assertThrows(RynBridgeError::class.java) {
            negotiator.parse("1.2")
        }
    }

    @Test
    fun `compatible same major stable`() {
        assertTrue(negotiator.isCompatible(local = "1.0.0", remote = "1.2.3"))
        assertTrue(negotiator.isCompatible(local = "2.0.0", remote = "2.5.0"))
    }

    @Test
    fun `incompatible different major stable`() {
        assertFalse(negotiator.isCompatible(local = "1.0.0", remote = "2.0.0"))
    }

    @Test
    fun `compatible same minor prerelease`() {
        assertTrue(negotiator.isCompatible(local = "0.1.0", remote = "0.1.5"))
    }

    @Test
    fun `incompatible different minor prerelease`() {
        assertFalse(negotiator.isCompatible(local = "0.1.0", remote = "0.2.0"))
    }

    @Test
    fun `assertCompatible throws`() {
        val exception = assertThrows(RynBridgeError::class.java) {
            negotiator.assertCompatible(local = "1.0.0", remote = "2.0.0")
        }
        assertEquals(ErrorCode.VERSION_MISMATCH, exception.code)
    }

    @Test
    fun `assertCompatible passes`() {
        assertDoesNotThrow {
            negotiator.assertCompatible(local = "0.1.0", remote = "0.1.3")
        }
    }

    @Test
    fun `invalid version not compatible`() {
        assertFalse(negotiator.isCompatible(local = "bad", remote = "1.0.0"))
    }
}
